torch/nn/modules/transformer.py

```diff
@@ -0,0 +1,242 @@
+	
+		for i in range(self.num_layers):
+			output = self.layers[i](output, memory, tgt_mask, memory_mask)
+
+		if self.norm:
+			output = self.norm(output)
+
+		return output
+
+class EncoderLayer(nn.Module):
+	"""
+	Encoder layer is made up of self-attn and feed forward (defined below)
+	Docs comming...
+	"""
+	def __init__(self, d_model, nhead, d_ff=2048,dropout=0.1):
+		super(EncoderLayer, self).__init__()
+		self.self_attn = MultiheadAttention(d_model, nhead, dropout=dropout)
+		self.ff = FeedForward(d_model, d_ff=d_ff, dropout=dropout)
+		self.norm1 = Tensor1DNorm(d_model)
+		self.norm2 = Tensor1DNorm(d_model)
+		self.dropout1 = nn.Dropout(dropout)
+		self.dropout2 = nn.Dropout(dropout)
+
+	def forward(self, x, mask):
+		x2 = self.norm1(x)
+		x = x + self.dropout1(self.self_attn(x2, x2, x2, mask=None))
+		x2 = self.norm2(x)
+		x = x + self.dropout2(self.ff(x2))
+		return x
+
+
+class DecoderLayer(nn.Module):
+	"""
+	Decoder layer is made up of self-attn, multihead_attn and feed forward (defined below)
+	Docs comming...
+	"""
+	def __init__(self, d_model, nhead, d_ff=2048, dropout=0.1):
+		super(DecoderLayer, self).__init__()
+		self.self_attn = MultiheadAttention(d_model, nhead, dropout=dropout)
+		self.multihead_attn = MultiheadAttention(d_model, nhead, dropout=dropout)
+		self.ff = FeedForward(d_model, d_ff=d_ff, dropout=dropout)
+		self.norm1 = Tensor1DNorm(d_model)
+		self.norm2 = Tensor1DNorm(d_model)
+		self.norm3 = Tensor1DNorm(d_model)
+		self.dropout1 = nn.Dropout(dropout)
+		self.dropout2 = nn.Dropout(dropout)
+		self.dropout3 = nn.Dropout(dropout)
+
+	def forward(self, x, memory, tgt_mask, src_mask):
+		x2 = self.norm1(x)
+		x = x + self.dropout1(self.self_attn(x2, x2, x2, tgt_mask))
+		x2 = self.norm2(x)
+		x = x + self.dropout2(self.multihead_attn(x2, memory, memory))
+		x2 = self.norm3(x)
+		x = x + self.dropout3(self.ff(x2))		
+		return x
+
+
+# Temporarily leave Tensor1DNorm module here. Will bemoved somewhere else.
+class Tensor1DNorm(nn.Module):
+	"Normalization component with two learnable variables"
+	def __init__(self, d_model, eps = 1e-6):
+		super(Tensor1DNorm, self).__init__()
+		self.size = d_model
+		# create two learnable parameters to calibrate normalisation
+		self.alpha = nn.Parameter(torch.ones(self.size))
+		self.bias = nn.Parameter(torch.zeros(self.size))
+		self.eps = eps
+	
+	def forward(self, x):
+		x_mean = x.mean(dim=-1, keepdim=True)
+		x_std = x.std(dim=-1, keepdim=True)
+		return self.alpha * (x - x_mean) / (x_std + self.eps) + self.bias
+
+# Temporarily leave FeedForward module here. Will be moved somewhere else.
+class FeedForward(nn.Module):
+	def __init__(self, d_model, d_ff=2048, dropout = 0.1):
+		super(FeedForward, self).__init__() 
+		# We set d_ff as a default to 2048
+		self.linear1 = nn.Linear(d_model, d_ff)
+		self.dropout = nn.Dropout(dropout)
+		self.linear2 = nn.Linear(d_ff, d_model)
+	
+	def forward(self, x):
+		return self.linear2(self.dropout(F.relu(self.linear1(x))))
```

Comment:
This is a very common pattern in general. For now I think it's ok if you do this explicitly instead of using a separate Module within Encoder and DecoderLayer and comment that it could be fused.